pkg/messaging/mqtt/pubsub_test.go

```diff
@@ -0,0 +1,287 @@
+// Copyright (c) Mainflux
+// SPDX-License-Identifier: Apache-2.0
+
+package mqtt_test
+
+import (
+	"fmt"
+	"testing"
+
+	"github.com/mainflux/mainflux/pkg/messaging"
+	"github.com/mainflux/mainflux/pkg/messaging/mqtt"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+const (
+	topic       = "topic"
+	chansPrefix = "channels"
+	channel     = "9b7b1b3f-b1b0-46a8-a717-b8213f9eda3b"
+	subtopic    = "engine"
+	clientID    = "9b7b1b3f-b1b0-46a8-a717-b8213f9eda3b"
+)
+
+var (
+	msgChan = make(chan messaging.Message)
+	data    = []byte("payload")
+)
+
+func TestPublisher(t *testing.T) {
+	// Subscribing with topic, and with subtopic, so that we can publish messages
+	err := pubsub.Subscribe(clientID, fmt.Sprintf("%s.%s", chansPrefix, topic), handler{false})
+	require.Nil(t, err, fmt.Sprintf("got unexpected error: %s", err))
+	err = pubsub.Subscribe(clientID, fmt.Sprintf("%s.%s.%s.", chansPrefix, topic, subtopic), handler{false})
+	require.Nil(t, err, fmt.Sprintf("got unexpected error: %s", err))
```

Comment:
I would prefer to use pure MQTT client for subscribing (`client.Subscribe`) directly instead of our Subscriber to isolate Pubslih testing from Subscriber. That is, we'll prevent errors in our Subscriber implementation to fail the Publish test. The same goes for our Subscriber - use native MQTT client to publish to prevent coupling between two. The same should be applied to all the other tests but in the separate PR.

Reply:
This being said, it would be nice to have one test that actually does test this compatibility between Publisher and Subscriber implementations so we're sure that messages are sent and received in a compatible way.

Reply:
I will implement this asap!

Reply:
I have implemented this.
Then, changed the TestPubSub to TestSubscriber

Then, created a new function -> TestPubSub, which needs to be simplified, but it basically does what TestPublisher did earlier, ie., use pubsub.Subscribe, so it's basically testing both Publisher and Subscriber together.

Reply:
`TestSubscriber` mixes testing `Subscribe` and `Unsubscribe` methods. TBH, I don't mind that because it's a part of the same functionality (even inseparable) but it's a little cleaner to have them separated. The complexity of executing test cases is an indication that we should split them. For `Subscribe` use the same principle as for the `Publish` test - use the plain client to publish messages you expect on the Subscriber side. For `Unsubscribe` - you obviously need to call `Subscribe` (you can do that before iterating through the cases, though), there is nothing we can do about it.